lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {CollectionKey, DeepRecord, Key, OnyxKey, KeyValueMapping, Selector} from './types';
+
+type NullableKeyValueMapping = {
+    [TKey in OnyxKey]: KeyValueMapping[TKey] | null;
+};
+
+type BaseConnectOptions<TKey extends OnyxKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, KeyValueMapping[TKey] | null>;
+};
+
+type ConnectOptions<TKey extends OnyxKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: Record<string, KeyValueMapping[TKey] | null> | null, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: KeyValueMapping[TKey] | null, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+type Collection<TKey extends Key, TMap, TValue> = {
+    [MapK in keyof TMap]: MapK extends `${TKey}${string}`
+        ? MapK extends `${TKey}`
+            ? never // forbids empty id
+            : TValue
+        : never;
+};
+
+type OnyxUpdate<TKey extends OnyxKey> =
+    | {
+          onyxMethod: typeof METHOD.SET;
+          key: TKey;
+          value: KeyValueMapping[TKey] | null;
+      }
+    | {
+          onyxMethod: typeof METHOD.MERGE;
+          key: TKey;
+          value: PartialDeep<KeyValueMapping[TKey]>;
+      }
+    | {
+          [TKey in CollectionKey]: {
+              onyxMethod: typeof METHOD.MERGE_COLLECTION;
+              key: TKey;
+              value: Record<`${TKey}${string}`, PartialDeep<KeyValueMapping[TKey]>>;
+          };
+      }[CollectionKey];
+
+type OnyxUpdates<TKeyList extends OnyxKey[]> = {
+    [TKey in keyof TKeyList]: OnyxUpdate<TKeyList[TKey]>;
+};
+
+type InitOptions = {
+    keys?: DeepRecord<string, string>;
+    initialKeyStates?: Partial<NullableKeyValueMapping>;
+    safeEvictionKeys?: Key[];
```

Comment:
Can be both regular and CollectionKeys https://github.com/Expensify/react-native-onyx/blob/8dba043803a6269e90148519a9bc26838e7dda1e/lib/Onyx.js#L1225-L1228
```ts
 safeEvictionKeys?: OnyxKey[];
```